docs/design/services/s3/transfermanager/prototype.java

```diff
@@ -0,0 +1,864 @@
+}
+
+/**
+ * Metrics for a completed transfer.
+ */
+public interface TransferMetrics {
+    /**
+     * The number of milliseconds that elapsed before this transfer completed.
+     */
+    long elapsedMillis();
+
+    /**
+     * The total number of bytes transferred.
+     */
+    long bytesTransferred();
+}
+
+
+/**
+ * A request to download an object. The object to download is specified using
+ * the {@link DownloadObjectSpecification} union type.
+ */
+public class DownloadObjectRequest extends AbstractTransferRequest {
+    /**
+     * The specification for how to download the object.
+     */
+    DownloadObjectSpecification downloadSpecification();
+
+    /**
+     * Known properties to help optimize an object download.
+     */
+    public Optional<KnownObjectProperties> knownProperties();
+
+    public static DownloadObjectRequest forPresignedUrl(URL presignedUrl) {
+        ...
+    }
+
+    public static UploadObjectRequest forBucketAndKey(String bucket, String key) {
+        ...
+    }
+
+    public interface Builder extends AbstractTransferRequest.Builder {
+        /**
+         * The specification for how to download the object.
+         */
+        Builder downloadSpecification(DownloadObjectSpecification downloadSpecification);
+
+       /**
+         * Set the override configuration for this request.
+         */
+        @Override
+        Builder overrideConfiguration(TransferOverrideConfiguration config);
+
+        /**
+         * Set the progress listeners for this request.
+         */
+        @Override
+        Builder progressListeners(Collection<TransferProgressListener> progressListeners);
+
+        /**
+         * Add an additional progress listener for this request, appending it to
+         * the list of currently configured listeners on this request.
+         */
+        @Override
+        Builder addProgressListener(TransferProgressListener progressListener);
+
+        Builder knownProperties(KnownObjectProperties knownProperties);
+
+        DownloadObjectRequest build();
+    }
+}
+
+/**
+ * Union type to contain the different ways to express an object download from
+ * S3.
+ */
+public class DownloadObjectSpecification {
+    /**
+     * Return this specification as a presigned URL.
+     *
+     * @throws IllegalStateException If this specifier is not a presigned URL.
+     */
+    URL asPresignedUrl();
+
+    /**
+     * Return this specification as a {@link GetObjectRequest API request}.
+     *
+     * @throws IllegalStateException If this specifier is not an API request.
+     */
+    GetObjectRequest asApiRequest();
+
+    /**
+     * Returns {@code true} if this is a presigned URL, {@code false} otherwise.
+     */
+    boolean isPresignedUrl();
+
+    /**
+     * Returns {@code true} if this is an API request, {@code false} otherwise.
+     */
+    boolean isApiRequest();
+
+    /**
+     * Visit this object.
+     */
+    void visit(DownloadObjectSpecificationVisitor visitor);
```

Comment:
Nit: If we end up not using this, we can probably exclude it. I don't know if a customer would want it.

Reply:
Removed